### PR TITLE
Replace Deposit Record with Event

### PR DESCRIPTION
**Pull request by waymobetta**
## Related Issue Or Context

Closes: #69 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

## Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have ensured that all acceptance criteria (or expected behavior) from issue are met
- [ ] I have updated the documentation locally and in chainbridge-docs.
- [ ] I have added tests to cover my changes.
- [ ] I have ensured that all the checks are passing and green, I've signed the CLA bot

### CODE REVIEW

File: chains/evm/listener/listener.go [hunk 1]
```diff
@@ -17,10 +17,23 @@ import (
...
+	// ID of chain deposit will be bridged to
 	DestinationID uint8
-	ResourceID    [32]byte
-	DepositNonce  uint64
+	// ResourceID used to find address of handler to be used for deposit
+	ResourceID [32]byte
+	// Nonce of deposit
+	DepositNonce uint64
+	// Address (msg.sender: user)
+	Address common.Address
```
Comment (P1sar) [inline chains/evm/listener/listener.go:30]: i think it will be more meaning-full to rename it to SenderAddress or something like that

File: chains/evm/listener/event-handler.go [hunk 2]
```diff
@@ -101,50 +100,24 @@ func toCallArg(msg ethereum.CallMsg) map[string]interface{} {
...
 		Source:       sourceID,
 		Destination:  destId,
 		DepositNonce: nonce,
-		ResourceId:   out0.ResourceID,
+		ResourceId:   resourceID,
 		Type:         relayer.FungibleTransfer,
 		Payload: []interface{}{
-			out0.Amount.Bytes(),
-			out0.DestinationRecipientAddress,
+			calldata,
```
Comment (P1sar) [inline chains/evm/listener/event-handler.go:118]: It should not work this way. Because check what ERC20MessageHandler expects to see in Payload, ther is 2 way, you can parse calldata here and pass expected data for messageHnalder or change how the handler behaves, id rather select second

**Review by P1sar — CHANGES_REQUESTED** [review]

File: chains/evm/listener/event-handler.go (hunk 2 above)
Comment (waymobetta) [reply chains/evm/listener/event-handler.go]: done

File: chains/evm/listener/event-handler.go [hunk 3]
```diff
@@ -101,50 +103,28 @@ func toCallArg(msg ethereum.CallMsg) map[string]interface{} {
...
-		return nil, err
-	}
-	res, err := a.Unpack("getDepositRecord", out)
-	if err != nil {
-		return nil, err
-	}
-	if len(res) == 0 {
-		return nil, errors.New("no handler associated with such resourceID")
-	}
+	// TODO: refactor
```
Comment (P1sar) [inline chains/evm/listener/event-handler.go:114]: why todo? seems ok solution. 
Everything else looks ok

File: chains/evm/listener/event-handler.go [hunk 4]
```diff
@@ -29,27 +28,28 @@ func NewETHEventHandler(address common.Address, client ChainClient) *ETHEventHan
...
 	}
 
-	return eventHandler(sourceID, destID, depositNonce, addr, e.client)
+	return eventHandler(sourceID, destID, depositNonce, resourceID, calldata, handlerResponse)
 }
 
-func (e *ETHEventHandler) matchResourceIDToHandlerAddress(rID [32]byte) (common.Address, error) {
+// matchResourceIDToHandlerAddress is a private method that matches a previously registered resource ID to its corresponding handler address
+func (e *ETHEventHandler) matchResourceIDToHandlerAddress(resourceID [32]byte) (common.Address, error) {
 	definition := "[{\"inputs\":[{\"internalType\":\"bytes32\",\"name\":\"\",\"type\":\"bytes32\"}],\"name\":\"_resourceIDToHandlerAddress\",\"outputs\":[{\"internalType\":\"address\",\"name\":\"\",\"type\":\"address\"}],\"stateMutability\":\"view\",\"type\":\"function\"}]"
```
Comment (mpetrun5) [inline chains/evm/listener/event-handler.go:47]: We should probably change now all ABIs to get definition from consts (or in separate PR)

File: chains/evm/listener/event-handler_test.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	sourceID := uint8(1)
+
+	message, err := Erc20EventHandler(sourceID, depositLog.DestinationID, depositLog.DepositNonce, depositLog.ResourceID, depositLog.Calldata, depositLog.HandlerResponse)
+	if err != nil {
+		t.Fatalf("could not generate event handler message: %v", err)
+	}
+
+	// TODO: refactor
```
Comment (mpetrun5) [inline chains/evm/listener/event-handler_test.go:41]: ?

File: chains/evm/listener/event-handler_test.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	"math/big"
+	"reflect"
+	"testing"
+
+	"github.com/ChainSafe/chainbridge-core/relayer"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/math"
+)
+
+func TestErc20EventHandler(t *testing.T) {
```
Comment (mpetrun5) [inline chains/evm/listener/event-handler_test.go:13]: We should probably be using test suites in tests to be consistent like https://github.com/ChainSafe/chainbridge-celo/blob/main/chain/listener/listener_test.go#L32

I think it should be probably called Erc20HandlerTestSuite to make it more easier to find tests associated with that code

File: chains/evm/listener/event-handler_test.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+		DepositNonce: depositLog.DepositNonce,
+		ResourceId:   depositLog.ResourceID,
+		Type:         relayer.FungibleTransfer,
+		Payload: []interface{}{
+			amountParsed,
+			recipientAddressParsed,
+		},
+	}
+
+	if !reflect.DeepEqual(message, expected) {
```
Comment (mpetrun5) [inline chains/evm/listener/event-handler_test.go:58]: s.Equal(foo, bar) when you implement test suite

File: chains/evm/listener/event-handler_test.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+		DepositNonce:    1,
+		SenderAddress:   common.HexToAddress("0x4CEEf6139f00F9F4535Ad19640Ff7A0137708485"),
+		Calldata:        calldata,
+		HandlerResponse: []byte{}, // empty []byte
+	}
+
+	sourceID := uint8(1)
+
+	message, err := Erc20EventHandler(sourceID, depositLog.DestinationID, depositLog.DepositNonce, depositLog.ResourceID, depositLog.Calldata, depositLog.HandlerResponse)
+	if err != nil {
```
Comment (mpetrun5) [inline chains/evm/listener/event-handler_test.go:37]: s.Nil(err)

File: chains/evm/listener/event-handler_test.go (hunk 6 above)
Comment (mpetrun5) [reply chains/evm/listener/event-handler_test.go:13]: Should probably also add test for when calldata len is zero to have 100% coverage :smile:

File: chains/evm/listener/event-handler.go [hunk 9]
```diff
@@ -15,7 +15,7 @@ import (
 )
 
 type EventHandlers map[common.Address]EventHandlerFunc
-type EventHandlerFunc func(sourceID, destId uint8, nonce uint64, handlerContractAddress common.Address, caller ChainClient) (*relayer.Message, error)
+type EventHandlerFunc func(sourceID, destId uint8, nonce uint64, resourceID [32]byte, calldata, handlerResponse []byte) (*relayer.Message, error)
```
Comment (mpetrun5) [inline chains/evm/listener/event-handler.go:18]:
```suggestion
type EventHandlerFunc func(sourceID, destId uint8, nonce uint64, resourceID internalTypes.ResourceID, calldata, handlerResponse []byte) (*relayer.Message, error)
```

File: chains/evm/listener/listener_test.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+	s.mockEventHandler.EXPECT().HandleEvent(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(expected, nil)
+
+	message, err := s.mockEventHandler.HandleEvent(sourceID, depositLog.DestinationID, depositLog.DepositNonce, depositLog.ResourceID, depositLog.Calldata, depositLog.HandlerResponse)
+
+	s.Nil(err)
+
+	s.NotNil(message)
+
+	if !s.Equal(message, expected) {
```
Comment (mpetrun5) [inline chains/evm/listener/listener_test.go:78]: I am pretty sure you just need s.Equal(message, expected) without the if.
it will fail automatically if it isn't the same

File: chains/evm/listener/event-handler.go [hunk 11]
```diff
@@ -102,50 +104,26 @@ func toCallArg(msg ethereum.CallMsg) map[string]interface{} {
...
-		return nil, err
-	}
-	res, err := a.Unpack("getDepositRecord", out)
-	if err != nil {
-		return nil, err
-	}
-	if len(res) == 0 {
-		return nil, errors.New("no handler associated with such resourceID")
-	}
+	amount := calldata[:32]
```
Comment (mpetrun5) [inline chains/evm/listener/event-handler.go:115]: Can it in theory happen that calldata length is something less than 32 and it panics?
